drivers/ieee802154/ieee802154_kw41z.c

```diff
@@ -425,130 +660,217 @@ static int kw41z_tx(struct device *dev, struct net_pkt *pkt,
+			"SEQ_STATE: 0x%08X, state: %d",
+			irqsts, (unsigned int)ZLL->PHY_CTRL,
+			(unsigned int)seq_state, state);
+
+	/* Clear interrupts */
+	ZLL->IRQSTS = irqsts;
 
-	/* TMR2 IRQ - time-out */
-	if ((irqsts & ZLL_IRQSTS_TMR2IRQ_MASK) &&
-	    !(irqsts & ZLL_IRQSTS_TMR2MSK_MASK)) {
-		SYS_LOG_DBG("TMR2 timeout");
-		kw41z_tmr2_disable();
-		atomic_set(&kw41z_context_data.seq_retval, -EBUSY);
-		k_sem_give(&kw41z_context_data.seq_sync);
+	if ((!(ZLL->PHY_CTRL & ZLL_PHY_CTRL_RX_WMRK_MSK_MASK)) &&
+	    (irqsts & ZLL_IRQSTS_RXWTRMRKIRQ_MASK)) {
+		/*
+		 * There is a bug in the KW41Z where in noisy environments
+		 * the RX sequence can get lost. The watermark mask IRQ can
+		 * start TMR3 to complete the rest of the read or to assert
+		 * IRQ if the sequencer gets lost so we can reset things.
+		 * Note that a TX from the upper layers will also reset
+		 * things so the problem is contained a bit in normal
+		 * operation.
+		 */
+		rx_len = (irqsts & ZLL_IRQSTS_RX_FRAME_LENGTH_MASK)
+			>> ZLL_IRQSTS_RX_FRAME_LENGTH_SHIFT;
+
+		SYS_LOG_DBG("WMRK irq: seq_state: 0x%08x, rx_len: %d",
+			seq_state, rx_len);
+
+		KW_DBG_TRACE(KW41_DBG_TRACE_WTRM, irqsts,
+			(unsigned int)ZLL->PHY_CTRL, seq_state);
+
+		/* Convert to symbols and add IFS and ACK duration */
+		rx_len = rx_len * 2 + 12 + 22 + 2;
+		kw41z_tmr3_set_timeout(rx_len);
+		restart_rx = 0;
 	}
 
 	/* Sequence done IRQ */
 	if ((state != KW41Z_STATE_IDLE) && (irqsts & ZLL_IRQSTS_SEQIRQ_MASK)) {
-		kw41z_disable_seq_irq();
-		kw41z_tmr1_disable();
-		kw41z_set_seq_state(KW41Z_STATE_IDLE);
-
-		switch (state) {
-		case KW41Z_STATE_RX:
-			SYS_LOG_DBG("RX seq done");
-
-			/*
-			 * KW41Z seems to require some time before the RX SEQ
-			 * done IRQ is triggered and the data is actually
-			 * available in the packet buffer.
-			 */
-			k_busy_wait(KW41Z_PRE_RX_WAIT_TIME);
-
-			rx_len = (ZLL->IRQSTS &
-				  ZLL_IRQSTS_RX_FRAME_LENGTH_MASK) >>
-				 ZLL_IRQSTS_RX_FRAME_LENGTH_SHIFT;
-			if (rx_len != 0) {
-				kw41z_rx(&kw41z_context_data, rx_len);
-			}
-
-			break;
-		case KW41Z_STATE_TXRX:
-			SYS_LOG_DBG("TXRX seq done");
-			kw41z_tmr2_disable();
-		case KW41Z_STATE_TX:
-			SYS_LOG_DBG("TX seq done");
-			if (irqsts & ZLL_IRQSTS_CCA_MASK) {
+		/*
+		 * PLL unlock, the autosequence has been aborted due to
+		 * PLL unlock
+		 */
+		if (irqsts & ZLL_IRQSTS_PLL_UNLOCK_IRQ_MASK) {
+			kw41z_isr_seq_cleanup();
+			restart_rx = 1;
+		}
+		/*
+		 * TMR3 timeout, the autosequence has been aborted due to
+		 * TMR3 timeout
+		 */
+		else if ((irqsts & ZLL_IRQSTS_TMR3IRQ_MASK) &&
+			(!(irqsts & ZLL_IRQSTS_RXIRQ_MASK)) &&
+			(state != KW41Z_STATE_TX)) {
+
+			SYS_LOG_INF("a) TMR3 timeout: irqsts: 0x%08X, "
+				"seq_state: 0x%08X, state: %d",
+				irqsts, seq_state, state);
+
+			KW_DBG_TRACE(KW41_DBG_TRACE_TMR3, irqsts,
+				(unsigned int)ZLL->PHY_CTRL, seq_state);
+
+			kw41z_isr_timeout_cleanup();
+			restart_rx = 1;
+			if (state == KW41Z_STATE_TXRX) {
```

Comment:
break a line  before if/while/etc... unless the previous assignment is the one tested in the condition.